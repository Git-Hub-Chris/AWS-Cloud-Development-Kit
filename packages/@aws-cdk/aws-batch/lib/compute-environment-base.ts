import * as iam from '@aws-cdk/aws-iam';
import { IResource, ITaggable, Resource, TagManager } from '@aws-cdk/core';
import { Construct } from 'constructs';


export interface IComputeEnvironment extends IResource, iam.IGrantable, ITaggable {
  /**
   * The name of the ComputeEnvironment
   *
   * @default - generated by CloudFormation
   */
  readonly name?: string;
  /**
   * The role Batch uses to perform actions on your behalf
   * in your account, such as provision instances to run your jobs
   *
   * @default - a serviceRole will be created for managed CEs, none for unmanaged CEs
   */
  readonly serviceRole?: iam.IRole;

  /**
   * Whether or not this ComputeEnvironment can accept jobs from a Queue.
   * Enabled ComputeEnvironments can accept jobs from a Queue and
   * can scale instances up or down.
   * Disabled ComputeEnvironments cannot accept jobs from a Queue or
   * scale instances up or down.
   *
   * If you change a ComputeEnvironment from enabled to disabled while it is executing jobs,
   * Jobs in the `STARTED` or `RUNNING` states will not
   * be interrupted. The ComputeEnvironment will no longer scale instances down;
   * that is, any instances that were scaled up will remain if the ComputeEnvironment is disabled.
   *
   * To ensure you aren't billed for unused capacity, remove the ComputeEnvironment from your stack.
   */
  readonly enabled?: boolean;
}

export interface ComputeEnvironmentProps {
  /**
 * The name of the ComputeEnvironment
 *
 * @default - generated by CloudFormation
 */
  readonly name?: string;
  /**
   * The role Batch uses to perform actions on your behalf
   * in your account, such as provision instances to run your jobs
   *
   * @default - a serviceRole will be created for managed CEs, none for unmanaged CEs
   */
  readonly serviceRole?: iam.IRole;

  /**
   * Whether or not this ComputeEnvironment can accept jobs from a Queue.
   * Enabled ComputeEnvironments can accept jobs from a Queue and
   * can scale instances up or down.
   * Disabled ComputeEnvironments cannot accept jobs from a Queue or
   * scale instances up or down.
   *
   * If you change a ComputeEnvironment from enabled to disabled while it is executing jobs,
   * Jobs in the `STARTED` or `RUNNING` states will not
   * be interrupted. The ComputeEnvironment will no longer scale instances down;
   * that is, any instances that were scaled up will remain if the ComputeEnvironment is disabled.
   *
   * To ensure you aren't billed for unused capacity, remove the ComputeEnvironment from your stack.
   */
  readonly enabled?: boolean;
}

export abstract class ComputeEnvironmentBase extends Resource implements IComputeEnvironment {
  name?: string | undefined;
  serviceRole?: iam.IRole | undefined;
  enabled?: boolean | undefined;
  grantPrincipal: iam.IPrincipal;
  tags: TagManager;

  constructor(scope: Construct, id: string, props: ComputeEnvironmentProps) {
    super(scope, id);

    this.name = props.name;
    this.serviceRole = props.serviceRole;
  }
}
