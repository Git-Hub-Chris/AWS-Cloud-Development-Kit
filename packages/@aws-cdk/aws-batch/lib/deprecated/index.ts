// Deprecated Batch L2
export * from './exposed-secret';
export * from './compute-environment';
export * from './job-definition';
export * from './job-queue';
