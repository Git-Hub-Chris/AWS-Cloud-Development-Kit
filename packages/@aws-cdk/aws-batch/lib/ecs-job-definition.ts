import { ArnFormat, Stack } from '@aws-cdk/core';
import { Construct } from 'constructs';
import { CfnJobDefinition } from './batch.generated';
import { EcsEc2ContainerDefinition, IEcsContainerDefinition } from './ecs-container-definition';
import { baseJobDefinitionProperties, IJobDefinition, JobDefinitionBase, JobDefinitionProps } from './job-definition-base';

/**
 * A JobDefinition that uses ECS orchestration
 */
interface IEcsJobDefinition extends IJobDefinition {
  /**
   * The container that this job will run
   */
  readonly container: IEcsContainerDefinition
}

/**
 * @internal
 */
export enum Compatibility {
  EC2 = 'EC2',
  FARGATE = 'FARGATE',
}

/**
 * Props for EcsJobDefinition
 */
export interface EcsJobDefinitionProps extends JobDefinitionProps {
  /**
   * The container that this job will run
   */
  readonly container: IEcsContainerDefinition
}

/**
 * A JobDefinition that uses ECS orchestration
 *
 * @resource AWS::Batch::JobDefinition
 */
export class EcsJobDefinition extends JobDefinitionBase implements IEcsJobDefinition {
  /**
   * Import a JobDefinition by its arn.
   */
  public static fromJobDefinitionArn(scope: Construct, id: string, jobDefinitionArn: string): IJobDefinition {
    const stack = Stack.of(scope);
    const jobDefinitionName = stack.splitArn(jobDefinitionArn, ArnFormat.SLASH_RESOURCE_NAME).resourceName!;

    class Import extends JobDefinitionBase implements IEcsJobDefinition {
      public readonly jobDefinitionArn = jobDefinitionArn;
      public readonly jobDefinitionName = jobDefinitionName;
      public readonly enabled = true;
      container = {} as any;
    }

    return new Import(scope, id);
  }

  readonly container: IEcsContainerDefinition

  public readonly jobDefinitionArn: string;
  public readonly jobDefinitionName: string;

  constructor(scope: Construct, id: string, props: EcsJobDefinitionProps) {
    super(scope, id, props);

    this.container = props.container;

    const resource = new CfnJobDefinition(this, 'Resource', {
      ...baseJobDefinitionProperties(this),
      type: 'container',
      jobDefinitionName: props.jobDefinitionName,
      containerProperties: this.container?._renderContainerDefinition(),
      platformCapabilities: this.renderPlatformCapabilities(),
    });

    this.jobDefinitionArn = this.getResourceArnAttribute(resource.ref, {
      service: 'batch',
      resource: 'job-definition',
      resourceName: this.physicalName,
    });
    this.jobDefinitionName = this.getResourceNameAttribute(resource.ref);
  }

  private renderPlatformCapabilities() {
    if (this.container instanceof EcsEc2ContainerDefinition) {
      return [Compatibility.EC2];
    }

    return [Compatibility.FARGATE];
  }
}
