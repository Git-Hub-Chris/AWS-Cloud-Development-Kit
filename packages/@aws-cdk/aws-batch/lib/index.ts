// AWS::Batch CloudFormation Resources:
export * from './batch.generated';
//export * from './deprecated';
export * from './managed-compute-environment';
export * from './ecs-job-definition';
//export { IJobDefinition, RetryStrategy, Reason, JobDefinitionProps } from './job-definition-base';
export * from './job-definition-base';
export { IComputeEnvironment, ComputeEnvironmentProps } from './compute-environment-base';
export * from './eks-job-definition';
export * from './multinode-job-definition';
export * from './ecs-container-definition';
export * from './eks-container-definition';
export * from './job-queue';
export * from './scheduling-policy';