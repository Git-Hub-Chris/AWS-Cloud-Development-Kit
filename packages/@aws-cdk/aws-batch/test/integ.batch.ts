import ec2 = require('@aws-cdk/aws-ec2');
import ecr = require('@aws-cdk/aws-ecr');
import { ContainerImage, EcrImage } from '@aws-cdk/aws-ecs';
import cdk = require('@aws-cdk/core');
import * as batch from '../lib/';

export const app = new cdk.App();

const stack = new cdk.Stack(app, 'batch-stack');

const vpc = new ec2.Vpc(stack, 'vpc');

new batch.JobQueue(stack, 'batch-job-queue', {
  computeEnvironmentOrder: [
    {
      computeEnvironment: new batch.ComputeEnvironment(stack, 'batch-managed-compute-env'),
      order: 1,
    },
    {
      computeEnvironment: new batch.ComputeEnvironment(stack, 'batch-demand-compute-env', {
        managed: false,
        computeResources: {
          type: batch.ComputeResourceType.ON_DEMAND,
          vpc,
        },
      }),
      order: 2,
    },
    {
      computeEnvironment: new batch.ComputeEnvironment(stack, 'batch-spot-compute-env', {
        managed: false,
        computeResources: {
          type: batch.ComputeResourceType.SPOT,
          vpc,
          bidPercentage: 80,
        },
      }),
      order: 3,
    },
  ]
});

const repo = new ecr.Repository(stack, 'batch-job-repo');

new batch.JobDefinition(stack, 'batch-job-def-from-ecr', {
  container: {
    image: new EcrImage(repo, 'latest'),
  },
});

new batch.JobDefinition(stack, 'batch-job-def-from-', {
  container: {
    image: ContainerImage.fromRegistry('docker/whalesay'),
  },
});
