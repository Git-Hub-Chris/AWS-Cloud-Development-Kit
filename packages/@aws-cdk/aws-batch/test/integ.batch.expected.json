{
  "Resources": {
    "vpcA2121C38": {
      "Type": "AWS::EC2::VPC",
      "Properties": {
        "CidrBlock": "10.0.0.0/16",
        "EnableDnsHostnames": true,
        "EnableDnsSupport": true,
        "InstanceTenancy": "default",
        "Tags": [
          {
            "Key": "Name",
            "Value": "batch-stack/vpc"
          }
        ]
      }
    },
    "vpcPublicSubnet1Subnet2E65531E": {
      "Type": "AWS::EC2::Subnet",
      "Properties": {
        "CidrBlock": "10.0.0.0/19",
        "VpcId": {
          "Ref": "vpcA2121C38"
        },
        "AvailabilityZone": "test-region-1a",
        "MapPublicIpOnLaunch": true,
        "Tags": [
          {
            "Key": "Name",
            "Value": "batch-stack/vpc/PublicSubnet1"
          },
          {
            "Key": "aws-cdk:subnet-name",
            "Value": "Public"
          },
          {
            "Key": "aws-cdk:subnet-type",
            "Value": "Public"
          }
        ]
      }
    },
    "vpcPublicSubnet1RouteTable48A2DF9B": {
      "Type": "AWS::EC2::RouteTable",
      "Properties": {
        "VpcId": {
          "Ref": "vpcA2121C38"
        },
        "Tags": [
          {
            "Key": "Name",
            "Value": "batch-stack/vpc/PublicSubnet1"
          }
        ]
      }
    },
    "vpcPublicSubnet1RouteTableAssociation5D3F4579": {
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
      "Properties": {
        "RouteTableId": {
          "Ref": "vpcPublicSubnet1RouteTable48A2DF9B"
        },
        "SubnetId": {
          "Ref": "vpcPublicSubnet1Subnet2E65531E"
        }
      }
    },
    "vpcPublicSubnet1DefaultRoute10708846": {
      "Type": "AWS::EC2::Route",
      "Properties": {
        "RouteTableId": {
          "Ref": "vpcPublicSubnet1RouteTable48A2DF9B"
        },
        "DestinationCidrBlock": "0.0.0.0/0",
        "GatewayId": {
          "Ref": "vpcIGWE57CBDCA"
        }
      },
      "DependsOn": [
        "vpcVPCGW7984C166"
      ]
    },
    "vpcPublicSubnet1EIPDA49DCBE": {
      "Type": "AWS::EC2::EIP",
      "Properties": {
        "Domain": "vpc"
      }
    },
    "vpcPublicSubnet1NATGateway9C16659E": {
      "Type": "AWS::EC2::NatGateway",
      "Properties": {
        "AllocationId": {
          "Fn::GetAtt": [
            "vpcPublicSubnet1EIPDA49DCBE",
            "AllocationId"
          ]
        },
        "SubnetId": {
          "Ref": "vpcPublicSubnet1Subnet2E65531E"
        },
        "Tags": [
          {
            "Key": "Name",
            "Value": "batch-stack/vpc/PublicSubnet1"
          }
        ]
      }
    },
    "vpcPublicSubnet2Subnet009B674F": {
      "Type": "AWS::EC2::Subnet",
      "Properties": {
        "CidrBlock": "10.0.32.0/19",
        "VpcId": {
          "Ref": "vpcA2121C38"
        },
        "AvailabilityZone": "test-region-1b",
        "MapPublicIpOnLaunch": true,
        "Tags": [
          {
            "Key": "Name",
            "Value": "batch-stack/vpc/PublicSubnet2"
          },
          {
            "Key": "aws-cdk:subnet-name",
            "Value": "Public"
          },
          {
            "Key": "aws-cdk:subnet-type",
            "Value": "Public"
          }
        ]
      }
    },
    "vpcPublicSubnet2RouteTableEB40D4CB": {
      "Type": "AWS::EC2::RouteTable",
      "Properties": {
        "VpcId": {
          "Ref": "vpcA2121C38"
        },
        "Tags": [
          {
            "Key": "Name",
            "Value": "batch-stack/vpc/PublicSubnet2"
          }
        ]
      }
    },
    "vpcPublicSubnet2RouteTableAssociation21F81B59": {
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
      "Properties": {
        "RouteTableId": {
          "Ref": "vpcPublicSubnet2RouteTableEB40D4CB"
        },
        "SubnetId": {
          "Ref": "vpcPublicSubnet2Subnet009B674F"
        }
      }
    },
    "vpcPublicSubnet2DefaultRouteA1EC0F60": {
      "Type": "AWS::EC2::Route",
      "Properties": {
        "RouteTableId": {
          "Ref": "vpcPublicSubnet2RouteTableEB40D4CB"
        },
        "DestinationCidrBlock": "0.0.0.0/0",
        "GatewayId": {
          "Ref": "vpcIGWE57CBDCA"
        }
      },
      "DependsOn": [
        "vpcVPCGW7984C166"
      ]
    },
    "vpcPublicSubnet2EIP9B3743B1": {
      "Type": "AWS::EC2::EIP",
      "Properties": {
        "Domain": "vpc"
      }
    },
    "vpcPublicSubnet2NATGateway9B8AE11A": {
      "Type": "AWS::EC2::NatGateway",
      "Properties": {
        "AllocationId": {
          "Fn::GetAtt": [
            "vpcPublicSubnet2EIP9B3743B1",
            "AllocationId"
          ]
        },
        "SubnetId": {
          "Ref": "vpcPublicSubnet2Subnet009B674F"
        },
        "Tags": [
          {
            "Key": "Name",
            "Value": "batch-stack/vpc/PublicSubnet2"
          }
        ]
      }
    },
    "vpcPublicSubnet3Subnet11B92D7C": {
      "Type": "AWS::EC2::Subnet",
      "Properties": {
        "CidrBlock": "10.0.64.0/19",
        "VpcId": {
          "Ref": "vpcA2121C38"
        },
        "AvailabilityZone": "test-region-1c",
        "MapPublicIpOnLaunch": true,
        "Tags": [
          {
            "Key": "Name",
            "Value": "batch-stack/vpc/PublicSubnet3"
          },
          {
            "Key": "aws-cdk:subnet-name",
            "Value": "Public"
          },
          {
            "Key": "aws-cdk:subnet-type",
            "Value": "Public"
          }
        ]
      }
    },
    "vpcPublicSubnet3RouteTableA3C00665": {
      "Type": "AWS::EC2::RouteTable",
      "Properties": {
        "VpcId": {
          "Ref": "vpcA2121C38"
        },
        "Tags": [
          {
            "Key": "Name",
            "Value": "batch-stack/vpc/PublicSubnet3"
          }
        ]
      }
    },
    "vpcPublicSubnet3RouteTableAssociationD102D1C4": {
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
      "Properties": {
        "RouteTableId": {
          "Ref": "vpcPublicSubnet3RouteTableA3C00665"
        },
        "SubnetId": {
          "Ref": "vpcPublicSubnet3Subnet11B92D7C"
        }
      }
    },
    "vpcPublicSubnet3DefaultRoute3F356A11": {
      "Type": "AWS::EC2::Route",
      "Properties": {
        "RouteTableId": {
          "Ref": "vpcPublicSubnet3RouteTableA3C00665"
        },
        "DestinationCidrBlock": "0.0.0.0/0",
        "GatewayId": {
          "Ref": "vpcIGWE57CBDCA"
        }
      },
      "DependsOn": [
        "vpcVPCGW7984C166"
      ]
    },
    "vpcPublicSubnet3EIP2C3B9D91": {
      "Type": "AWS::EC2::EIP",
      "Properties": {
        "Domain": "vpc"
      }
    },
    "vpcPublicSubnet3NATGateway82F6CA9E": {
      "Type": "AWS::EC2::NatGateway",
      "Properties": {
        "AllocationId": {
          "Fn::GetAtt": [
            "vpcPublicSubnet3EIP2C3B9D91",
            "AllocationId"
          ]
        },
        "SubnetId": {
          "Ref": "vpcPublicSubnet3Subnet11B92D7C"
        },
        "Tags": [
          {
            "Key": "Name",
            "Value": "batch-stack/vpc/PublicSubnet3"
          }
        ]
      }
    },
    "vpcPrivateSubnet1Subnet934893E8": {
      "Type": "AWS::EC2::Subnet",
      "Properties": {
        "CidrBlock": "10.0.96.0/19",
        "VpcId": {
          "Ref": "vpcA2121C38"
        },
        "AvailabilityZone": "test-region-1a",
        "MapPublicIpOnLaunch": false,
        "Tags": [
          {
            "Key": "Name",
            "Value": "batch-stack/vpc/PrivateSubnet1"
          },
          {
            "Key": "aws-cdk:subnet-name",
            "Value": "Private"
          },
          {
            "Key": "aws-cdk:subnet-type",
            "Value": "Private"
          }
        ]
      }
    },
    "vpcPrivateSubnet1RouteTableB41A48CC": {
      "Type": "AWS::EC2::RouteTable",
      "Properties": {
        "VpcId": {
          "Ref": "vpcA2121C38"
        },
        "Tags": [
          {
            "Key": "Name",
            "Value": "batch-stack/vpc/PrivateSubnet1"
          }
        ]
      }
    },
    "vpcPrivateSubnet1RouteTableAssociation67945127": {
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
      "Properties": {
        "RouteTableId": {
          "Ref": "vpcPrivateSubnet1RouteTableB41A48CC"
        },
        "SubnetId": {
          "Ref": "vpcPrivateSubnet1Subnet934893E8"
        }
      }
    },
    "vpcPrivateSubnet1DefaultRoute1AA8E2E5": {
      "Type": "AWS::EC2::Route",
      "Properties": {
        "RouteTableId": {
          "Ref": "vpcPrivateSubnet1RouteTableB41A48CC"
        },
        "DestinationCidrBlock": "0.0.0.0/0",
        "NatGatewayId": {
          "Ref": "vpcPublicSubnet1NATGateway9C16659E"
        }
      }
    },
    "vpcPrivateSubnet2Subnet7031C2BA": {
      "Type": "AWS::EC2::Subnet",
      "Properties": {
        "CidrBlock": "10.0.128.0/19",
        "VpcId": {
          "Ref": "vpcA2121C38"
        },
        "AvailabilityZone": "test-region-1b",
        "MapPublicIpOnLaunch": false,
        "Tags": [
          {
            "Key": "Name",
            "Value": "batch-stack/vpc/PrivateSubnet2"
          },
          {
            "Key": "aws-cdk:subnet-name",
            "Value": "Private"
          },
          {
            "Key": "aws-cdk:subnet-type",
            "Value": "Private"
          }
        ]
      }
    },
    "vpcPrivateSubnet2RouteTable7280F23E": {
      "Type": "AWS::EC2::RouteTable",
      "Properties": {
        "VpcId": {
          "Ref": "vpcA2121C38"
        },
        "Tags": [
          {
            "Key": "Name",
            "Value": "batch-stack/vpc/PrivateSubnet2"
          }
        ]
      }
    },
    "vpcPrivateSubnet2RouteTableAssociation007E94D3": {
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
      "Properties": {
        "RouteTableId": {
          "Ref": "vpcPrivateSubnet2RouteTable7280F23E"
        },
        "SubnetId": {
          "Ref": "vpcPrivateSubnet2Subnet7031C2BA"
        }
      }
    },
    "vpcPrivateSubnet2DefaultRouteB0E07F99": {
      "Type": "AWS::EC2::Route",
      "Properties": {
        "RouteTableId": {
          "Ref": "vpcPrivateSubnet2RouteTable7280F23E"
        },
        "DestinationCidrBlock": "0.0.0.0/0",
        "NatGatewayId": {
          "Ref": "vpcPublicSubnet2NATGateway9B8AE11A"
        }
      }
    },
    "vpcPrivateSubnet3Subnet985AC459": {
      "Type": "AWS::EC2::Subnet",
      "Properties": {
        "CidrBlock": "10.0.160.0/19",
        "VpcId": {
          "Ref": "vpcA2121C38"
        },
        "AvailabilityZone": "test-region-1c",
        "MapPublicIpOnLaunch": false,
        "Tags": [
          {
            "Key": "Name",
            "Value": "batch-stack/vpc/PrivateSubnet3"
          },
          {
            "Key": "aws-cdk:subnet-name",
            "Value": "Private"
          },
          {
            "Key": "aws-cdk:subnet-type",
            "Value": "Private"
          }
        ]
      }
    },
    "vpcPrivateSubnet3RouteTable24DA79A0": {
      "Type": "AWS::EC2::RouteTable",
      "Properties": {
        "VpcId": {
          "Ref": "vpcA2121C38"
        },
        "Tags": [
          {
            "Key": "Name",
            "Value": "batch-stack/vpc/PrivateSubnet3"
          }
        ]
      }
    },
    "vpcPrivateSubnet3RouteTableAssociationC58B3C2C": {
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
      "Properties": {
        "RouteTableId": {
          "Ref": "vpcPrivateSubnet3RouteTable24DA79A0"
        },
        "SubnetId": {
          "Ref": "vpcPrivateSubnet3Subnet985AC459"
        }
      }
    },
    "vpcPrivateSubnet3DefaultRoute30C45F47": {
      "Type": "AWS::EC2::Route",
      "Properties": {
        "RouteTableId": {
          "Ref": "vpcPrivateSubnet3RouteTable24DA79A0"
        },
        "DestinationCidrBlock": "0.0.0.0/0",
        "NatGatewayId": {
          "Ref": "vpcPublicSubnet3NATGateway82F6CA9E"
        }
      }
    },
    "vpcIGWE57CBDCA": {
      "Type": "AWS::EC2::InternetGateway",
      "Properties": {
        "Tags": [
          {
            "Key": "Name",
            "Value": "batch-stack/vpc"
          }
        ]
      }
    },
    "vpcVPCGW7984C166": {
      "Type": "AWS::EC2::VPCGatewayAttachment",
      "Properties": {
        "VpcId": {
          "Ref": "vpcA2121C38"
        },
        "InternetGatewayId": {
          "Ref": "vpcIGWE57CBDCA"
        }
      }
    },
    "batchdemandcomputeenvResourceInstanceRole8989496E": {
      "Type": "AWS::IAM::Role",
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "batch.amazonaws.com"
              }
            }
          ],
          "Version": "2012-10-17"
        }
      }
    },
    "batchdemandcomputeenvResourceServiceInstanceRole8DF7CB96": {
      "Type": "AWS::IAM::Role",
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "batch.amazonaws.com"
              }
            }
          ],
          "Version": "2012-10-17"
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition"
                },
                ":iam::aws:policy/service-role/AWSBatchServiceRole"
              ]
            ]
          }
        ]
      }
    },
    "batchdemandcomputeenv6131030A": {
      "Type": "AWS::Batch::ComputeEnvironment",
      "Properties": {
        "ServiceRole": {
          "Fn::GetAtt": [
            "batchdemandcomputeenvResourceServiceInstanceRole8DF7CB96",
            "Arn"
          ]
        },
        "Type": "UNMANAGED",
        "ComputeResources": {
          "AllocationStrategy": "BEST_FIT",
          "InstanceRole": {
            "Fn::GetAtt": [
              "batchdemandcomputeenvResourceInstanceRole8989496E",
              "Arn"
            ]
          },
          "InstanceTypes": [
            "optimal"
          ],
          "MaxvCpus": 256,
          "MinvCpus": 0,
          "Subnets": [
            {
              "Ref": "vpcPrivateSubnet1Subnet934893E8"
            },
            {
              "Ref": "vpcPrivateSubnet2Subnet7031C2BA"
            },
            {
              "Ref": "vpcPrivateSubnet3Subnet985AC459"
            }
          ],
          "Type": "EC2"
        },
        "State": "ENABLED"
      }
    },
    "batchspotcomputeenvResourceInstanceRoleF6188F15": {
      "Type": "AWS::IAM::Role",
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "batch.amazonaws.com"
              }
            }
          ],
          "Version": "2012-10-17"
        }
      }
    },
    "batchspotcomputeenvResourceServiceInstanceRole8B0DF5A7": {
      "Type": "AWS::IAM::Role",
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "batch.amazonaws.com"
              }
            }
          ],
          "Version": "2012-10-17"
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition"
                },
                ":iam::aws:policy/service-role/AWSBatchServiceRole"
              ]
            ]
          }
        ]
      }
    },
    "batchspotcomputeenv2CE4DFD9": {
      "Type": "AWS::Batch::ComputeEnvironment",
      "Properties": {
        "ServiceRole": {
          "Fn::GetAtt": [
            "batchspotcomputeenvResourceServiceInstanceRole8B0DF5A7",
            "Arn"
          ]
        },
        "Type": "UNMANAGED",
        "ComputeResources": {
          "AllocationStrategy": "BEST_FIT",
          "BidPercentage": 80,
          "InstanceRole": {
            "Fn::GetAtt": [
              "batchspotcomputeenvResourceInstanceRoleF6188F15",
              "Arn"
            ]
          },
          "InstanceTypes": [
            "optimal"
          ],
          "MaxvCpus": 256,
          "MinvCpus": 0,
          "Subnets": [
            {
              "Ref": "vpcPrivateSubnet1Subnet934893E8"
            },
            {
              "Ref": "vpcPrivateSubnet2Subnet7031C2BA"
            },
            {
              "Ref": "vpcPrivateSubnet3Subnet985AC459"
            }
          ],
          "Type": "SPOT"
        },
        "State": "ENABLED"
      }
    },
    "batchjobqueueE3C528F2": {
      "Type": "AWS::Batch::JobQueue",
      "Properties": {
        "ComputeEnvironmentOrder": [
          {
            "ComputeEnvironment": {
              "Ref": "batchdemandcomputeenv6131030A"
            },
            "Order": 1
          },
          {
            "ComputeEnvironment": {
              "Ref": "batchspotcomputeenv2CE4DFD9"
            },
            "Order": 2
          }
        ],
        "Priority": 1,
        "State": "ENABLED"
      }
    },
    "batchjobrepo4C508C51": {
      "Type": "AWS::ECR::Repository",
      "UpdateReplacePolicy": "Retain",
      "DeletionPolicy": "Retain"
    },
    "batchjobdeffromecrE0E30DAD": {
      "Type": "AWS::Batch::JobDefinition",
      "Properties": {
        "Type": "container",
        "ContainerProperties": {
          "Image": {
            "Fn::Join": [
              "",
              [
                {
                  "Fn::Select": [
                    4,
                    {
                      "Fn::Split": [
                        ":",
                        {
                          "Fn::GetAtt": [
                            "batchjobrepo4C508C51",
                            "Arn"
                          ]
                        }
                      ]
                    }
                  ]
                },
                ".dkr.ecr.",
                {
                  "Fn::Select": [
                    3,
                    {
                      "Fn::Split": [
                        ":",
                        {
                          "Fn::GetAtt": [
                            "batchjobrepo4C508C51",
                            "Arn"
                          ]
                        }
                      ]
                    }
                  ]
                },
                ".",
                {
                  "Ref": "AWS::URLSuffix"
                },
                "/",
                {
                  "Ref": "batchjobrepo4C508C51"
                },
                ":latest"
              ]
            ]
          },
          "Memory": 4,
          "Privileged": false,
          "ReadonlyRootFilesystem": false,
          "Vcpus": 1
        },
        "RetryStrategy": {
          "Attempts": 1
        },
        "Timeout": {}
      }
    },
    "batchjobdeffrom4007378D": {
      "Type": "AWS::Batch::JobDefinition",
      "Properties": {
        "Type": "container",
        "ContainerProperties": {
          "Image": "docker/whalesay",
          "Memory": 4,
          "Privileged": false,
          "ReadonlyRootFilesystem": false,
          "Vcpus": 1
        },
        "RetryStrategy": {
          "Attempts": 1
        },
        "Timeout": {}
      }
    }
  }
}